eth/fetcher/tx_fetcher.go

```diff
@@ -0,0 +1,623 @@
+				if time.Since(instance)+gatherSlack > arriveTimeout {
+					// Transaction expired without propagation, schedule for retrieval
+					if f.announced[hash] != nil {
+						panic("announce tracker already contains waitlist item")
+					}
+					f.announced[hash] = f.waitlist[hash]
+					for peer := range f.waitlist[hash] {
+						if announces := f.announces[peer]; announces != nil {
+							announces[hash] = struct{}{}
+						} else {
+							f.announces[peer] = map[common.Hash]struct{}{hash: struct{}{}}
+						}
+						delete(f.waitslots[peer], hash)
+						if len(f.waitslots[peer]) == 0 {
+							delete(f.waitslots, peer)
+						}
+						actives[peer] = struct{}{}
+					}
+					delete(f.waittime, hash)
+					delete(f.waitlist, hash)
+				}
+			}
+			// If transactions are still waiting for propagation, reschedule the wait timer
+			if len(f.waittime) > 0 {
+				f.rescheduleWait(waitTrigger)
+			}
+			// If any peers became active and are idle, request transactions from them
+			if len(actives) > 0 {
+				f.scheduleFetches(timeoutTrigger, actives)
+			}
+
+		case <-timeoutTrigger:
+			// Clean up any expired retrievals and avoid re-requesting them from the
+			// same peer (either overloaded or malicious, useless in both cases). We
+			// could also penalize (Drop), but there's nothing to gain, and if could
+			// possibly further increase the load on it.
+			for peer, req := range f.requests {
+				if time.Since(req.time)+gatherSlack > txFetchTimeout {
+					// Reschedule all the not-yet-delivered fetches to alternate peers
+					delete(f.requests, peer)
+					for _, hash := range req.hashes {
+						if _, ok := f.announced[hash]; ok {
+							panic("announced tracker already contains alternate item")
+						}
+						if f.alternates[hash] != nil { // nil if tx was broadcast during fetch
+							f.announced[hash] = f.alternates[hash]
+						}
+						delete(f.alternates, hash)
+						delete(f.fetching, hash)
+					}
+					txFetchTimeoutMeter.Mark(int64(len(req.hashes)))
+				}
+			}
+			// Schedule a new transaction retrieval
+			f.scheduleFetches(timeoutTrigger, nil)
+
+		case delivery := <-f.cleanup:
+			// If the request failed, reschedule all the hashes
+			if delivery.direct && delivery.failed {
+				req := f.requests[delivery.origin]
+				if req == nil {
+					log.Warn("Unexpected transaction delivery failure", "peer", delivery.origin)
+					break
+				}
+				delete(f.requests, delivery.origin)
+
+				for _, hash := range req.hashes {
+					if _, ok := f.announced[hash]; ok {
+						panic("announced tracker already contains alternate item")
+					}
+					f.announced[hash] = f.alternates[hash]
+					delete(f.alternates, hash)
+					delete(f.fetching, hash)
+				}
+				break
+			}
+			// Independent if the delivery was direct or broadcast, remove all
+			// traces of the hash from internal trackers
+			for _, hash := range delivery.hashes {
+				if _, ok := f.waitlist[hash]; ok {
+					for peer, txset := range f.waitslots {
```

Comment:
For every delivered tx, we iterate for every peer, txset in waitslots. So, what if you request `H1, H2`, and I deliver `T1, T1, T1, T1, ... T2`. IIUC, `hashes` will contain `H1, H1, H1 ... H2`. That will cause a bit of spinning here, but maybe that's not an issue.?